src/nvim/lua/executor.c

```diff
@@ -624,9 +676,61 @@ void nlua_init(void)
   luaL_openlibs(lstate);
   nlua_state_init(lstate);
 
+  luv_set_thread_cb(nlua_thread_acquire_vm, NULL);
+
   global_lstate = lstate;
 }
 
+static lua_State *nlua_thread_acquire_vm(void)
+{
+  lua_State *lstate = luaL_newstate();
+
+  // Add in the lua standard libraries
+  luaL_openlibs(lstate);
+
+  // print
+  lua_pushcfunction(lstate, &nlua_print);
+  lua_setglobal(lstate, "print");
+
+  // vim
+  lua_newtable(lstate);
+  // vim.is_therad
+  lua_pushboolean(lstate, true);
+  lua_setfield(lstate, LUA_REGISTRYINDEX, "nvim.thread");
+  lua_pushcfunction(lstate, &nlua_is_thread);
+  lua_setfield(lstate, -2, "is_thread");
+
+  // vim.loop
+  luv_set_callback(lstate, nlua_luv_thread_cfpcall);
+  luaopen_luv(lstate);
+  lua_pushvalue(lstate, -1);
+  lua_setfield(lstate, -3, "loop");
+
+  lua_getglobal(lstate, "package");
+  lua_getfield(lstate, -1, "loaded");
+  lua_pushvalue(lstate, -3);
+  lua_setfield(lstate, -2, "luv");
+  lua_pop(lstate, 3);
+
+  // vim.mpack
```

Comment:
`vim.NIL` and `vim._empty_dict_mt` is also needed for this. Though I'll suggest breaking out a helper function for shared libs (mpack, diff, also `shared.lua` ?) and move these there instead of duplicating more code.

Reply:
Done.